xdsl/dialects/experimental/dmp.py

```diff
@@ -426,6 +427,160 @@ def print_parameters(self, printer: Printer) -> None:
         printer.print_string(">")
 
 
+class DomainDecompositionStrategy(ParametrizedAttribute, ABC):
+    def calc_resize(self, shape: tuple[int, ...]) -> tuple[int, ...]:
+        raise NotImplementedError("SlicingStrategy must implement calc_resize!")
+
+    def halo_exchange_defs(self, shape: ShapeAttr) -> Iterable[ExchangeDeclarationAttr]:
+        raise NotImplementedError("SlicingStrategy must implement halo_exchange_defs!")
+
+    def comm_layout(self) -> RankTopoAttr:
+        raise NotImplementedError("SlicingStrategy must implement comm_count!")
+
+
+@irdl_attr_definition
+class GridSlice2dAttr(DomainDecompositionStrategy):
+    """
+    Takes a grid with two or more dimensions, slices it along the first two into equally
+    sized segments.
+    """
+
+    name = "dmp.grid_slice_2d"
+
+    topology: ParameterDef[RankTopoAttr]
+
+    diagonals: ParameterDef[builtin.BoolAttr]
+
+    def __init__(self, topo: tuple[int, ...]):
+        super().__init__(
+            [RankTopoAttr(topo), builtin.BoolAttr.from_int_and_width(0, 1)]
+        )
+
+    def _verify(self):
+        assert (
+            len(self.topology.as_tuple()) >= 2
+        ), "GridSlice2d requires at least two dimensions"
+
+    def calc_resize(self, shape: tuple[int, ...]) -> tuple[int, ...]:
+        assert len(shape) >= 2, "GridSlice2d requires at least two dimensions"
+        for size, node_count in zip(shape, self.topology.as_tuple()):
+            assert (
+                size % node_count == 0
+            ), "GridSlice2d requires domain be neatly divisible by shape"
+        return (
+            *(
+                size // node_count
+                for size, node_count in zip(shape, self.topology.as_tuple())
+            ),
+            *(size for size in shape[2:]),
+        )
+
+    def halo_exchange_defs(self, shape: ShapeAttr) -> Iterable[ExchangeDeclarationAttr]:
+        yield from _flat_face_exchanges_for_dim(shape, 0)
+
+        yield from _flat_face_exchanges_for_dim(shape, 1)
+
+        # TOOD: add diagonals
```

Comment:
TOOD

Reply:
But also probably better to `raise NotImplementedError("Diagonals not yet implemented")`